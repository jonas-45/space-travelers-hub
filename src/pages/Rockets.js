import React from 'react';
import { useSelector } from 'react-redux';
import RocketItem from '../components/RocketItem';

const Rockets = () => {
  const { rockets } = useSelector((state) => state.rocket);
  const isLoading = useSelector((state) => state.rocket.isLoading);
  return (
    <div>
      {(isLoading && (
      <div style={{
        display: 'flex', justifyContent: 'center', alignItems: 'center', width: '100%', height: '100vh', paddingBottom: '20%',
      }}
      >
        <div className="spinner-border text-warning" role="status">
          <span className="sr-only" />
        </div>
      </div>
      ))}
      {!isLoading && (rockets.map((rocketObj) => (
        <RocketItem rocket={rocketObj} key={rocketObj.id} />
      )))}
    </div>
  );
};

export default Rockets;
