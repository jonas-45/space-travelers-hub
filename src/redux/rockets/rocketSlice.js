import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

const url = 'https://api.spacexdata.com/v3/rockets';

export const getRockets = createAsyncThunk('rockets/data', async () => {
  console.log('Starting fetching');
  const response = await fetch(url).then((resp) => resp.json()).catch((error) => console.log(error));
  const rocketsData = await response.map((rocketObj) => ({
    id: rocketObj.rocket_id,
    name: rocketObj.rocket_name,
    type: rocketObj.rocket_type,
    describe: rocketObj.description,
    flickr_images: rocketObj.flickr_images[0],
    reserved: false,
  }));
  
  return rocketsData;
});

const initialState = {
  rockets: [],
};

const rocketSlice = createSlice({
  name: 'rockets',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(getRockets.fulfilled, (state, action) => ({
      ...state,
      rockets: action.payload,
    }));
  },
});

export default rocketSlice.reducer;
