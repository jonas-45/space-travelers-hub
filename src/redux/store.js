import { configureStore } from '@reduxjs/toolkit';
import rocketReducer from './rockets/rocketSlice';
import missionsReducer from './missions/missionsSlice';

const store = configureStore({
  reducer: {
    missions: missionsReducer,
    rocket: rocketReducer,
  },
});

export default store;
