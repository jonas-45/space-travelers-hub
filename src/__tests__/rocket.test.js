import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import axios from 'axios';
import Rockets from '../pages/Rockets';
import store from '../redux/store';
import { getRockets, reserveRocket } from '../redux/rockets/rocketSlice';

it('Rocket page renders correctly', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Rockets />
      </Provider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

// Rocket slicer and reducer check
describe('Rocket redux state tests', () => {
  it('Should initially hold Rocket data by default', () => {
    const state = store.getState().rocket;
    expect(state.rockets.length).toEqual(0);
  });

  it('should Reserve Rocket payload send correct', () => {
    const expectedPayload = { payload: '9D1B7E0', type: 'rockets/reserveRocket' };
    const actualPayload = reserveRocket('9D1B7E0');
    expect(actualPayload).toEqual(expectedPayload);
  });
  it('Rockets fetch data from API', async () => {
    const url = 'https://api.spacexdata.com/v3/rockets';
    const axiosSpy = jest.spyOn(axios, 'get');
    jest.setTimeout(90000);
    const dispatchSpy = jest.fn();
    await getRockets(url)(dispatchSpy);
    expect(axiosSpy).toHaveBeenCalledWith(url);
  });
});
