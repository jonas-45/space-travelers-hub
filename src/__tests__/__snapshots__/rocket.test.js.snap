// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rocket page renders correctly 1`] = `
<div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "height": "100vh",
        "justifyContent": "center",
        "paddingBottom": "20%",
        "width": "100%",
      }
    }
  >
    <div
      className="spinner-border text-warning"
      role="status"
    >
      <span
        className="sr-only"
      />
    </div>
  </div>
</div>
`;
